File: python/random_forest.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jun 27 11:26:46 2025
@author: USER
"""

# 🌲 Random Forest Classification

# =============================
# 1. Import Required Libraries
# =============================
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix, accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from matplotlib.colors import ListedColormap

# =============================
# 2. Load Dataset
# =============================
dataset = pd.read_csv('../data/Social_network_ads.csv')
X = dataset.iloc[:, :-1].values  # Age and EstimatedSalary
y = dataset.iloc[:, -1].values   # Purchased

# =============================
# 3. Split Dataset
# =============================
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# =============================
# 4. Feature Scaling
# =============================
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# =============================
# 5. Train Random Forest Classifier
# =============================
classifier = RandomForestClassifier(n_estimators=10, random_state=0)
classifier.fit(X_train, y_train)

# =============================
# 6. Predictions and Evaluation
# =============================
y_pred = classifier.predict(X_test)

cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:\n", cm)
print("Accuracy:", accuracy_score(y_test, y_pred))

sns.heatmap(cm, annot=True, fmt='d', cmap='Blues', cbar=False)
plt.title("Confusion Matrix")
plt.xlabel("Predicted")
plt.ylabel("Actual")
plt.tight_layout()
plt.show()

# =============================
# 7. Visualization Function
# =============================
def plot_decision_boundary(X_set, y_set, title):
    X1, X2 = np.meshgrid(
        np.arange(X_set[:, 0].min() - 1, X_set[:, 0].max() + 1, 0.01),
        np.arange(X_set[:, 1].min() - 1, X_set[:, 1].max() + 1, 0.01)
    )
    Z = classifier.predict(np.c_[X1.ravel(), X2.ravel()]).reshape(X1.shape)

    plt.contourf(X1, X2, Z, alpha=0.25, cmap=ListedColormap(('red', 'green')))
    plt.xlim(X1.min(), X1.max())
    plt.ylim(X2.min(), X2.max())

    for i, j in enumerate(np.unique(y_set)):
        plt.scatter(
            X_set[y_set == j, 0], X_set[y_set == j, 1],
            color=('red', 'green')[i], label=f"Class {j}"
        )
    
    plt.xlabel('Age')
    plt.ylabel('Estimated Salary')
    plt.title(title)
    plt.legend()
    plt.tight_layout()
    plt.show()

# =============================
# 8. Visualize Training Set Results
# =============================
plot_decision_boundary(X_train, y_train, "Random Forest Classifier - Training Set")

# =============================
# 9. Visualize Test Set Results
# =============================
plot_decision_boundary(X_test, y_test, "Random Forest Classifier - Test Set")

File: r/random_forest.R
# Random Forest Classification

# Load dataset
dataset <- read.csv('../data/Social_network_ads.csv')

# Split dataset into training and test sets
library(caTools)
set.seed(123)
split <- sample.split(dataset$Purchased, SplitRatio = 0.8)
dataset_train <- subset(dataset, split == TRUE)
dataset_test  <- subset(dataset, split == FALSE)

# Feature Scaling
dataset_train[,-3] <- scale(dataset_train[,-3])
dataset_test[,-3]  <- scale(dataset_test[,-3])

# Fit Random Forest model
library(randomForest)
classifier <- randomForest(
  x = dataset_train[,-3],
  y = factor(dataset_train$Purchased),
  ntree = 10
)

# Predict on test set
y_pred <- predict(classifier, newdata = dataset_test[,-3])

# Confusion Matrix
cm <- table(Actual = dataset_test$Purchased, Predicted = y_pred)
print(cm)

# Visualize decision boundary - Training Set
set <- dataset_train
X1 <- seq(min(set[,1]) - 1, max(set[,1]) + 1, by = 0.01)
X2 <- seq(min(set[,2]) - 1, max(set[,2]) + 1, by = 0.01)
grid_set <- expand.grid(Age = X1, EstimatedSalary = X2)
y_grid <- predict(classifier, newdata = grid_set)

# Ensure plot area is large enough
par(mfrow = c(1, 1))
par(mar = c(4, 4, 2, 1))

plot(set[,-3],
     main = 'Random Forest (Training Set)',
     xlab = 'Age',
     ylab = 'Estimated Salary',
     xlim = range(X1),
     ylim = range(X2))
contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
points(grid_set, pch = '.', col = ifelse(y_grid == 1, 'springgreen3', 'tomato'))
points(set, pch = 21, bg = ifelse(set[,3] == 1, 'green4', 'red3'))

# Visualize decision boundary - Test Set
set <- dataset_test
X1 <- seq(min(set[,1]) - 1, max(set[,1]) + 1, by = 0.01)
X2 <- seq(min(set[,2]) - 1, max(set[,2]) + 1, by = 0.01)
grid_set <- expand.grid(Age = X1, EstimatedSalary = X2)
y_grid <- predict(classifier, newdata = grid_set)

plot(set[,-3],
     main = 'Random Forest (Test Set)',
     xlab = 'Age',
     ylab = 'Estimated Salary',
     xlim = range(X1),
     ylim = range(X2))
contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
points(grid_set, pch = '.', col = ifelse(y_grid == 1, 'springgreen3', 'tomato'))
points(set, pch = 21, bg = ifelse(set[,3] == 1, 'green4', 'red3'))
